theano/sandbox/cuda/nnet.py

```diff
@@ -739,3 +739,124 @@ def c_support_code_apply(self, node, nodename):
+        - `iSize` is the size of each of these input blocks.
+        - `iWin` is the number of blocks that will be used as inputs. Which blocks
+          will be used is specified in `inputIdx`.
+        - `oBlocks` is the number or possible output blocks.
+        - `oSize` is the size of each of these output blocks.
+        - `oWin` is the number of output blocks that will actually be computed.
+          Which blocks will be computed is specified in `outputIdx`.
+        """
+        o = theano.tensor.as_tensor_variable(o)
+        W = theano.tensor.as_tensor_variable(W)
+        h = theano.tensor.as_tensor_variable(h)
+        inputIdx = theano.tensor.as_tensor_variable(inputIdx)
+        outputIdx = theano.tensor.as_tensor_variable(outputIdx)
+
+        if o.ndim != 3:
+            raise TypeError('The output o must be a 3D tensor')
+        if W.ndim != 4:
+            raise TypeError('The weight matrix W must be a 4D tensor')
+        if h.ndim != 3:
+            raise TypeError('The input h must be a 3D tensor')
+        if inputIdx.ndim != 2:
+            raise TypeError('The input indices inputIdx must be a 2D tensor')
+        if outputIdx.ndim != 2:
+            raise TypeError('The output indices outputIdx must be a 2D tensor')
+
+        assert inputIdx.type.dtype in discrete_dtypes
+        assert outputIdx.type.dtype in discrete_dtypes
+
+        return Apply(self, [o, W, h, inputIdx, outputIdx], [o.type()])
+
+    def perform(self, node, inp, out_):
+        raise NotImplementedError('Optimization of SparseBlock failed.')
+
+
+def sparse_block_dot_cpu(o, W, h, inputIdx, outputIdx):
+    """
+    Creates a graph for the sparse block dot operation. Check SparseBlockDot's
+    docstring for information about the arguments.
+    """
+    def _outer_loop_over_outputIdx(i, W, h, inputIdx, outputIdx):
+
+        def _inner_loop_over_inputIdx(j, W, h, inputIdx, outputIdx):
+
+            return T.dot(h[:, j, :], W[inputIdx[j], outputIdx[i], outputIdx[i], :, :])
+
+        a = theano.scan(fn=_inner_loop_over_inputIdx,
+                        sequences=T.arange(0, inputIdx.shape[1]),
+                        non_sequences=[W, h, inputIdx, outputIdx],
+                        name='_inner_loop_over_inputIdx')[0]
+        return a.sum()
+
+    b = theano.scan(fn=_outer_loop_over_outputIdx,
+                    sequences=T.arange(0, outputIdx.shape[1]),
+                    non_sequences=[W, h, inputIdx, outputIdx],
+                    name='_outer_loop_over_outputIdx')[0]
+
+    return T.set_subtensor(o[:, :, :], b.dimshuffle(1, 0, 2))
+
+
+@local_optimizer([SparseBlockDot])
+def local_sparse_block_dot_device(node):
+    is_running_on_gpu = False
+    for inp in node.inputs:
+        if isinstance(inp, CudaNdarrayType):
+            is_running_on_gpu = True
+            break
+
+    if is_running_on_gpu:
+        return [sparse_block_gemv_ss(*node.inputs)] # Todo: the name of this op should include gpu. What does gemv_ss stand for?
```

Comment:
It stand for **ge**neral **m**atrix-**v**ector multiplication. ```ss``` stand for **s**parse **s**parse because @abergeron was planning to implement a version where one matrix is sparse and the other is dense. I would agree to add gpu in the name, but gemv and ss should stay there.